chore - linting changes

devinea opened the pull request:
Changes 
- `pnpm pretty-quick --staged --ignore-path .prettierignore` to 
`pnpm pretty-quick --ignore-path .prettierignore` using alias `pnpm run:pretty-quick` as many developers commit without staging.
- Run linting on test folder (static checks and ensures prettier has been applied to the code)
   - remove `test` from `.eslintignore`
   - add `test` to  `tesconfig.json`
   - created `tsconfig-build.json` to exclude `test` from the build command so that test is not emitted to `dist`
   - Apply fixes from lint
- Apply prettier formatting. 

Verified `eslint-disable-next-line` statements or corrections. Each module needs two people as verifiers.

- [x] axios-extension @ullasholla @tobiasqueck 
- [ ] environment-check @tobiasqueck  
- [x] fiori-elements-writer @unseen1980 @IainSAP 
- [x] fiori-freestyle-writer @unseen1980 @IainSAP 
- [ ] logger - @tobiasqueck 
- [ ] store - @tobiasqueck 
- [ ] ui5-config @tobiasqueck 
- [x] ui5-proxy-middleware - @tobiasqueck @zdravko-georgiev 

Following modules don't have changes in `src` or `test`

- [x] backend-proxy-middleware
- [x] btp-utils
- [x] fe-fpm-writer (#699)
- [x] odata-service-writer (#702)
- [x] ui-components (#700)
- [x] ui5-application-writer (#700)
- [x] ui5-test-writer
- [x] yaml

devinea commented:
> I have reviewed all modified `.ts` files and there are now JSDoc comments generated for functions that don't contain any information. This needs to be fixed by the corresponding code owners.

@tobiasqueck Since they are in test files it probably best to disable jsdoc rules for test files?

tobiasqueck commented:
> > I have reviewed all modified `.ts` files and there are now JSDoc comments generated for functions that don't contain any information. This needs to be fixed by the corresponding code owners.
> 
> @tobiasqueck Since they are in test files it probably best to disable jsdoc rules for test files?

the ones I looked at in `axios-extension` and one of the middlewares were not test files. Otherwise, yes, agree, JSDoc is not required for the `.test.ts` files.

devinea commented:
> I have reviewed all modified `.ts` files and there are now JSDoc comments generated for functions that don't contain any information. This needs to be fixed by the corresponding code owners.

@tobiasqueck I have fixed most linting issues 
Most common remaining is `@typescript-eslint/no-non-null-assertion` https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-non-null-assertion.md
I am torn if we should "fix" the issues report or disable the rule. Any opinion? 

@Jimmy-Joseph19 There are 14 lint issues in ui-components remaining. Do you have time to look into them?
![Screenshot 2022-09-22 at 14 07 25](https://user-images.githubusercontent.com/10102054/191755225-bd7b0790-e29b-40ac-bc24-f821eb6c134b.png)

Jimmy-Joseph19 commented:
> > I have reviewed all modified `.ts` files and there are now JSDoc comments generated for functions that don't contain any information. This needs to be fixed by the corresponding code owners.
> 
> @tobiasqueck I have fixed most linting issues Most common remaining is `@typescript-eslint/no-non-null-assertion` https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-non-null-assertion.md I am torn if we should "fix" the issues report or disable the rule. Any opinion?
> 
> @Jimmy-Joseph19 There are 14 lint issues in ui-components remaining. Do you have time to look into them? ![Screenshot 2022-09-22 at 14 07 25](https://user-images.githubusercontent.com/10102054/191755225-bd7b0790-e29b-40ac-bc24-f821eb6c134b.png)

I can do it. Shall i push in same PR or different. Let me know thanks

Klaus-Keller commented:
@devinea, I find it very hard to spot changes to rules or to fix linting versus format changes. Can we split this up into two parts:

1. PR that does prettier format changes ONLY
2. PR that adjusts and fixes linting rules

devinea commented:
Thanks @Jimmy-Joseph19  for https://github.com/SAP/open-ux-tools/pull/689/commits/86189e40e6686aab9a595830d100ba1d9ccdc7b2

devinea commented:
> > Most common remaining is @typescript-eslint/no-non-null-assertion https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-non-null-assertion.md
> > I am torn if we should "fix" the issues report or disable the rule. Any opinion?
> 
> I actually like that rule. I think the `!` postfix operator is pretty dangerous if too lightly used.

OK I won't disable the rule. I adjusted code or annotated the line in some cases in https://github.com/SAP/open-ux-tools/pull/689/commits/8f4d04b74a5af0cee190a12d842d9dc18b81570b

devinea commented:
> @devinea, I find it very hard to spot changes to rules or to fix linting versus format changes. Can we split this up into two parts:
> 
> 1. PR that does prettier format changes ONLY
> 2. PR that adjusts and fixes linting rules

@Klaus-Keller Not sure that split would be that easy. Any chance you review commits as the commit message gives an idea of the changes included?

IainSAP reviewed (approved):
OK for `fiori-elements-writer`.

IainSAP commented on packages/fiori-freestyle-writer/test/common.ts:
```diff
@@ -9,6 +10,9 @@ export const testOutputDir = join(__dirname, '/test-output');
 
 export const debug = prepareDebug();
 
+/**
```
Empty jsdocs?

devinea replied:
fixed in https://github.com/SAP/open-ux-tools/pull/689/commits/028c6ad39d339a95190a55c617b6f19ebf72348b

IainSAP commented on examples/odata-cli/src/activities.ts:
```diff
@@ -20,6 +20,7 @@ export async function useCatalogAndFetchSomeMetadata(
         const atoSettings = await provider.getAtoInfo();
         if (!atoSettings || Object.keys(atoSettings).length === 0) {
+            // eslint-disable-next-line no-console
             console.warn('ATO setting is empty!');
```
Should be using the logging module?

devinea replied:
This is an example. So console should be ok?

ullasholla replied:
1 vote to leave the `console.warn` as it is. 
It's an example, let's keep it simple and only use the code that this is example for

IainSAP commented on examples/odata-cli/src/activities.ts:
```diff
@@ -45,6 +46,7 @@ export async function useCatalogAndFetchSomeMetadata(
         }
     } catch (error) {
+        // eslint-disable-next-line no-console
         console.error(error.cause || error.toString() || error);
```
Should be using the logging module?

devinea replied:
This is an example. So console should be ok?

ullasholla replied:
Agree with @devinea. Example code, let's keep it simple and focussed

Klaus-Keller commented:
> @Klaus-Keller Not sure that split would be that easy. Any chance you can review commits as the commit message gives an idea of the changes included?

@devinea, my reading issues are related to `.eslintrc` and figuring out what has really changed and what was formatting. I think we should make this obvious so everyone knows what rules are new or changed, even if they are just for test folders. So I'll share it here:

![image](https://user-images.githubusercontent.com/66327622/191800684-6b65126f-981f-461f-9e6d-c35902ec6379.png)

IainSAP reviewed (commented):
A general comment. The proliferation of eslint disables now indicates that we should modify some of these rules or re-implement the code. OK for test code but not published code. e.g. There are console logs in places that should be replaced?

Having separate rules for test code and published code might be appropriate.

devinea commented on .eslintrc:
```diff
@@ -77,12 +101,35 @@
+                ]
+            }
+        },
+        {
+            "files": [
+                "**/test/**/*.js",
+                "**/test/**/*.ts",
+                "**/test/**/*.tsx"
+            ],
+            "rules": {
+                "no-console": "off",
+                "jsdoc/require-returns-description": "off",
+                "jsdoc/require-param-description": "off",
+                "max-nested-callbacks": [
+                    "error",
+                    5
+                ],
+                "@typescript-eslint/no-non-null-assertion" : "off"
```
This is the new overrides to disable some rules on the test code.

ullasholla replied:
Do we have a lot of non-null assertions in test code?

devinea replied:
Some, but probably don't need to disable the rule.  
I removed the rule override and adjusted the code where appropriate. 
https://github.com/SAP/open-ux-tools/pull/689/commits/e33874490b876aa1139fcb5780fcfb7dd624156e

devinea commented:
> @devinea, my reading issues are related to `.eslintrc` and figuring out what has really changed and what was formatting. I think we should make this obvious so everyone knows what rules are new or changed, even if they are just for test folders. So I'll share it here:

Thanks @Klaus-Keller

ullasholla commented:
@Klaus-Keller:
> @devinea, my reading issues are related to `.eslintrc` and figuring out what has really changed and what was formatting. I think we should make this obvious so everyone knows what rules are new or changed, even if they are just for test folders. So I'll share it here:

Thanks for highlighting the relevant rules.

I find whitespace only changes distracting when there are lots of them.

Have you already tried hiding the whitespace only changes?
<img width="674" alt="image" src="https://user-images.githubusercontent.com/212316/191851087-28a64d80-0886-4052-b3c5-a60fca8307f1.png">

ullasholla commented on .eslintrc:
```diff
@@ -77,12 +101,35 @@
+                "jsdoc/require-param-description": "off",
+                "max-nested-callbacks": [
+                    "error",
+                    5
```
Is there a reason to pick `5` specifically? This is a warning in productive code, why make it an error here? This is especially relevant to test code because using nested callback like functions is the norm

devinea replied:
Good catch. Changed to warning in https://github.com/SAP/open-ux-tools/pull/689/commits/e33874490b876aa1139fcb5780fcfb7dd624156e 
The default is 3 and it raised a warning on many expect statements in the tests so I increased to 5 for test code

Klaus-Keller replied:
@ullasholla problem here lies in the nature of the structure of jest unit tests. A usual setup might look like

```
describe('Test description', () => {
    it('Should ...', () => {
        // Actual test
    });
    ..
});
```

As you can see, the actual test code is already enclosed in 2 levels of callbacks. 

My vote is also to increase setting for rule `max-nested-callbacks` by 2 levels for test code.

Klaus-Keller commented:
@ullasholla 
>Have you already tried hiding the whitespace only changes?

Yes, I did that but the formatting includes a lot of line breaks.

Klaus-Keller commented:
I reviewed module `@sap-ux/environment-check` and pushed some changes (https://github.com/SAP/open-ux-tools/pull/689/commits/4e62833ae11195840acf0ea3d1ec3cfb22b0d0ec).

While working on this I think I found conflicting eslint rules.

Following example:

![image](https://user-images.githubusercontent.com/66327622/191862470-65361433-25cc-4158-9e9b-61719ee6d73c.png)


It tells me, that I should quote the property `default`, as it is a reserved word. The rule comes from `quote-props` -> `as-needed` ->  `keywords: true`

https://github.com/SAP/open-ux-tools/blob/83fea4a611e43a646cbd454f3382d15d491c2062/.eslintrc#L214-L221

If I change it to quoted, I get another error:

![image](https://user-images.githubusercontent.com/66327622/191863042-95c3cafa-527a-4afb-b64a-64a306e81f21.png)

that tells me to replace `'default'` with `default`. This is caused by rule `prettier/prettier`:

https://github.com/SAP/open-ux-tools/blob/83fea4a611e43a646cbd454f3382d15d491c2062/.eslintrc#L255-L260

What helped was adding `"quoteProps": "preserve"`. The value `as-needed` didn't do the trick, I assume it does not consider keywords, like quote-props does.

Here is the final result with changed `prettier/prettier`, I hope this change is ok:

```
"prettier/prettier": [
            "error",
            {
                "endOfLine": "auto",
                "quoteProps": "preserve"
            }
        ]
```


By the way @devinea, when I use prettier to format `.eslintrc` I get the format as it is in `main` branch. If I format it in VSCode using `JSON Language Features` I get the same format as on branch `linting_changes`.

Any ideas what is correct formatting?

devinea commented:
> Here is the final result with changed `prettier/prettier`, I hope this change is ok:

Thanks @Klaus-Keller  for fixing that. I also changed `quoteProps` https://github.com/SAP/open-ux-tools/pull/689/commits/152f20c651a0ca5d2d483e57a9103da75b83f94e#diff-7312296367dd7c725b9beb500d7dab7bbeeb2aa03a1010a3fb2935f826d03fa6R19 as lint -fix was still removing the qoutes.

> By the way @devinea, when I use prettier to format .eslintrc I get the format as it is in main branch. If I format it in VSCode using JSON Language Features I get the same format as on branch linting_changes.

You are right, my bad. To fix I applied prettier formatting to the whole project in https://github.com/SAP/open-ux-tools/pull/689/commits/bb34775aaa8a1ce652556b51dfd767b7914ef88f 
added an excluded for `pnpm-lock.yaml` as I think prettier has been changing formatting of that file based on the pre-commit hook for some developers recently.

devinea commented:
> A general comment. The proliferation of eslint disables now indicates that we should modify some of these rules or re-implement the code. OK for test code but not published code. e.g. There are console logs in places that should be replaced?
> 
> Having separate rules for test code and published code might be appropriate.

I am open to opinions on this. I am OK with either. 
The approach I took was:
Change code to be compliant where it was a simple change. 
When it seems risker or changing the code didn't add value (e.g just more verbose) I add disable annotation. 
The benefit of keeping the rule it forces for new code the developer and reviewers to consider if it is OK to use that approach. 
If the rule is disabled it may not be considered as carefully.

Klaus-Keller commented:
>I am open to opinions on this. I am OK with either.
The approach I took was:
Change code to be compliant where it was a simple change.
When it seems risker or changing the code didn't add value (e.g just more verbose) I add disable annotation.
The benefit of keeping the rule it forces for new code the developer and reviewers to consider if it is OK to use that approach.
If the rule is disabled it may not be considered as carefully.

@devinea, what is the exact focus of this PR? Perhaps it got a too wide scope?

When I currently do
```
git clone https://github.com/SAP/open-ux-tools.git
cd open-ux-tools
pnpm install && pnpm build && pnpm lint
```

I see linting issues in multiple files across multiple modules. I think these issues should be addressed in separate PRs by respective module owners and should not be part of this PR. I'd split this PR into several smaller PRs:

1. Update prettier rules
2. Add linting rules for `test` folders and  enable it
3. Align tests to use UX_DEBUG consistently to write tests and clean folders

tobiasqueck commented on .changeset/warm-bees-tie.md:
```diff
@@ -0,0 +1,20 @@
+'@sap-ux/backend-proxy-middleware': patch
+'@sap-ux/btp-utils': patch
+'@sap-ux/environment-check': patch
+'@sap-ux/fe-fpm-writer': patch
```
Do we need to patch them? The `fe-fpm-writer` e.g. has only a change in the linting config, so it shouldn't affect the generated module.

devinea replied:
Yeah should no longer be needed.

Klaus-Keller commented on packages/fiori-freestyle-writer/test/common.ts:
```diff
@@ -9,10 +10,16 @@ export const testOutputDir = join(__dirname, '/test-output');
-    // eslint-disable-next-line no-console
+
     if (debug) {
+        // eslint-disable-next-line no-console
```
Perhaps we can agree, that usage of `console` is ok for `test` folder or avoid it everywhere.

devinea replied:
Good catch. Actually already allowed now in test folder rules. 
Remove disabled `no-console` in https://github.com/SAP/open-ux-tools/pull/689/commits/582c871523276087e806a7b80ddbb88a81111a27

tobiasqueck commented:
@zdravko-georgiev which modules did you review? Or did you review and approve all of them?

zdravko-georgiev commented:
> @zdravko-georgiev which modules did you review? Or did you review and approve all of them?

@tobiasqueck Sorry for not providing more information. I reviewed the proxy modules.

tobiasqueck commented on packages/environment-check/test/output/__snapshots__/markdown.test.ts.snap:
```diff
@@ -59,7 +59,6 @@ Dev Space Type: \`SAP Fiori\`<br>
     \\"zlib\\": \\"1.2.11\\",
     \\"ares\\": \\"1.17.2\\",
     \\"modules\\": \\"72\\",
-    \\"http_parser\\": \\"2.9.4\\",
```
What caused this change in the snapshot?

unseen1980 commented on packages/axios-extension/test/abap/ui5-abap-repository-service.test.ts:
```diff
@@ -172,6 +173,9 @@ describe('Ui5AbapRepositoryService', () => {
         test('ensure special characters are encoded', async () => {
+            /**
+             *
+             */
```
?

devinea replied:
Yeah this is an oddity, it causes an error to remove it.  
Would need to disable jsdoc completely for test files to remove it. 
Maybe it is the best option anyway?

tobiasqueck replied:
https://github.com/SAP/open-ux-tools/pull/689/commits/ced70fe6223b7c74af5523aa11f9aab255950af6

tobiasqueck commented:
@unseen1980 which modules did you review?

unseen1980 commented:
> @unseen1980 which modules did you review?

Mostly the fiori writers but I had a look at the rest of the files too. I didn't test the changes locally though. A couple of things that I didn't know and I had to google, such as ```root0```!

ullasholla reviewed (approved):
Reviewed changes in `axios-extension`. Made a few small changes to JSDoc comments. Thanks @devinea

Sonar check fails because the test coverage is below 80%. Makes sense to take care of the test coverage in other PRs?

zdravko-georgiev commented on packages/ui5-proxy-middleware/src/ui5/middleware.ts:
```diff
@@ -37,10 +37,11 @@ function createProxyOptions(logger: ToolsLogger, config: UI5ProxyConfig): Option
         for (const route of routes) {
             if (req.path.startsWith(route.route)) {
-                return route.handler(req, res, next);
+                route.handler(req, res, next);
+                return;
```
Why this change? A `RequestHander` should be returned, now it returns `undefined`.

tobiasqueck replied:
I does not need to return it, it just needs to call it. `createRequestHandler` returns a `RequestHandler` that checks the list of routes, and if it matches, it handles the request otherwise, it calls `next()`.
Previously the function was implicitly returning `RequestHandler | undefined` now we are explicit and made the function `void` because no one is reading the return value.

zdravko-georgiev reviewed (approved):
Reviewed proxy modules and logger

IainSAP reviewed (approved):
Thx @devinea , LGTM

- No unexpected changed in `@sap-ux/fiori-elements-writer`
- No unexpected changed in `@sap-ux/fiori-freestyle-writer`

tobiasqueck reviewed (approved):
I reviewed the modules that didn't have a reviewer as well as `axios-extension` and did not see any issues.
Further improvements could be made but this can happen in module specific PRs.

devinea commented:
Thanks All for commits, inputs and reviews. 
Merging as admin as the sonar code coverage is a little skewed as there is only 6 uncovered lines of code. 
Overall coverage is still good.
